File: Official Windows Platform Sample/Windows 8.1 Store app samples/[C#]-Windows 8.1 Store app samples/Print settings and print notifications/C# and JavaScript/DeviceAppForPrinters/js/print-background-task.js
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.
//
// Copyright (c) Microsoft Corporation. All rights reserved

//
// A JavaScript background task runs a specified JavaScript file.
//
(function () {
    "use strict";

    //
    // Save the printer name and asyncUI xml
    //
    var keyPrinterName = "BA5857FA-DE2C-4A4A-BEF2-49D8B4130A39";
    var keyAsyncUIXML = "55DCA47A-BEE9-43EB-A7C8-92ECA2FA0685";
    var settings = Windows.Storage.ApplicationData.current.localSettings;

    //
    // The background task instance's activation parameters are available via Windows.UI.WebUI.WebUIBackgroundTaskInstance.current
    //
    var backgroundTaskInstance = Windows.UI.WebUI.WebUIBackgroundTaskInstance.current;

    var details = backgroundTaskInstance.triggerDetails;
    settings.values[keyPrinterName] = details.printerName;
    settings.values[keyAsyncUIXML] = details.eventData;
    
    // With the print notification event details we can choose to show a toast, update the tile, or update the badge.
    // It is not recommended to always show a toast, especially for non-actionable events, as it may become annoying for most users.
    // User may even just turn off all toasts from this app, which is not a desired outcome.
    // For events that does not require user's immediate attention, it is recommended to update the tile/badge and not show a toast.
    showToast(details.printerName, details.eventData);
    updateTile(details.printerName, details.eventData);
    updateBadge();

    //
    // A JavaScript background task must call close when it is done.
    //
    close();

    function updateTile(printerName, bidiMessage) {
        var tileUpdater = Windows.UI.Notifications.TileUpdateManager.createTileUpdaterForApplication();
        tileUpdater.clear();

        var tileXml = Windows.UI.Notifications.TileUpdateManager.getTemplateContent(Windows.UI.Notifications.TileTemplateType.tileWide310x150Text09);
        var tileTextAttributes = tileXml.getElementsByTagName("text");
        tileTextAttributes[0].innerText = printerName;
        tileTextAttributes[1].innerText = bidiMessage;

        var tileNotification = new Windows.UI.Notifications.TileNotification(tileXml);
        tileNotification.Tag = "tag01";
        tileUpdater.update(tileNotification);
    }

    function updateBadge()
    {
        var badgeXml = Windows.UI.Notifications.BadgeUpdateManager.getTemplateContent(Windows.UI.Notifications.BadgeTemplateType.badgeGlyph);
        var badgeElement = badgeXml.selectSingleNode("/badge");
        badgeElement.setAttribute("value", "error");

        var badgeNotification = new Windows.UI.Notifications.BadgeNotification(badgeXml);
        Windows.UI.Notifications.BadgeUpdateManager.createBadgeUpdaterForApplication().update(badgeNotification);
    }

    function showToast(title, body) {
        var notifications = Windows.UI.Notifications;
        var toastNotificationManager = Windows.UI.Notifications.ToastNotificationManager;
        var toastXml = toastNotificationManager.getTemplateContent(notifications.ToastTemplateType.toastText02);

        var temp = "\"Printer Name\"" + ":" + "\"" + title + "\"" + "," + "\"AsyncUI XML\"" + ":" + "\"" + body + "\"";
        if (temp.length > 251) {
            temp = temp.substring(0, 251);
        }
        toastXml.selectSingleNode("/toast").setAttribute("launch", "'{" + temp + "}'");

        var textNodes = toastXml.getElementsByTagName("text");
        textNodes[0].appendChild(toastXml.createTextNode(title));
        textNodes[1].appendChild(toastXml.createTextNode(body));

        var toast = new notifications.ToastNotification(toastXml);
        toastNotificationManager.createToastNotifier().show(toast);
    }
})();

File: Official Windows Platform Sample/Windows 8 app samples/[C++]-Windows 8 app samples/C++/Windows 8 app samples/Device app for printers SDK sample (Windows 8)/C++/BackgroundTask/PrintBackgroundTask.cpp
#include "pch.h"
#include "PrintBackgroundTask.h"

using namespace BackgroundTask;
using namespace Platform;
using namespace Windows::UI::Notifications;
using namespace Windows::Data::Xml::Dom;

PrintBackgroundTask::PrintBackgroundTask(void)
{
    _keyPrinterName = ref new Platform::String(L"BA5857FA-DE2C-4A4A-BEF2-49D8B4130A39");
    _keyAsyncUIXML = ref new Platform::String(L"55DCA47A-BEE9-43EB-A7C8-92ECA2FA0685");
}

PrintBackgroundTask::~PrintBackgroundTask(void)
{
}

void PrintBackgroundTask::Run(Windows::ApplicationModel::Background::IBackgroundTaskInstance^ taskInstance)
{
    // Retrieve the printer name and event data
    Windows::Devices::Printers::Extensions::PrintNotificationEventDetails^ details = (Windows::Devices::Printers::Extensions::PrintNotificationEventDetails^)(taskInstance->TriggerDetails);
    Windows::Storage::ApplicationData^ applicationData = Windows::Storage::ApplicationData::Current;
    Windows::Foundation::Collections::IPropertySet^ propertySet = applicationData->LocalSettings->Values;
    propertySet->Remove(_keyPrinterName);
    propertySet->Remove(_keyAsyncUIXML);
    propertySet->Insert(_keyPrinterName, details->PrinterName);
    propertySet->Insert(_keyAsyncUIXML, details->EventData);

    // With the print notification event details we can choose to show a toast, update the tile, or update the badge.
    // It is not recommended to always show a toast, especially for non-actionable events, as it may become annoying for most users.
    // User may even just turn off all toasts from this app, which is not a desired outcome.
    // For events that does not require user's immediate attention, it is recommended to update the tile/badge and not show a toast.
    ShowToast(details->PrinterName, details->EventData);
    UpdateTile(details->PrinterName, details->EventData);
    UpdateBadge();
}

void PrintBackgroundTask::ShowToast(Platform::String^ title, Platform::String^ body)
{
    XmlDocument^ toastXml = ToastNotificationManager::GetTemplateContent(ToastTemplateType::ToastText02);

    //
    // The ToastText02 template has 2 text nodes (a header and a body)
    // Assign title to the first one, and body to the second one
    //
    XmlNodeList^ textList = toastXml->GetElementsByTagName(L"text");
    textList->Item(0)->AppendChild(toastXml->CreateTextNode(title));
    textList->Item(1)->AppendChild(toastXml->CreateTextNode(body));

    //
    // Set launch flag so the app can tell if it is being launched by the toast or from the tile
    //
    IXmlNode^ node = toastXml->SelectSingleNode(L"/toast");
    XmlElement^ element = (XmlElement^) node;
    element->SetAttribute(L"launch", title);

    //
    // Show the Toast
    //
    ToastNotification^ toast = ref new ToastNotification(toastXml);
    ToastNotificationManager::CreateToastNotifier()->Show(toast);
}

void PrintBackgroundTask::UpdateTile(Platform::String^ printerName, Platform::String^ bidiMessage)
{
    TileUpdater^ tileUpdater = TileUpdateManager::CreateTileUpdaterForApplication();
    if (nullptr != tileUpdater)
    {
        tileUpdater->Clear();

        XmlDocument^ tileXml = TileUpdateManager::GetTemplateContent(TileTemplateType::TileWideText09);
        if (nullptr != tileXml)
        {
            // Fill in the template
            XmlNodeList^ tileTextAttributes = tileXml->GetElementsByTagName("text");
            tileTextAttributes->GetAt(0)->InnerText = printerName;
            tileTextAttributes->GetAt(1)->InnerText = bidiMessage;

            // Create a tile update
            TileNotification^ tileNotification = ref new TileNotification(tileXml);
            if (nullptr != tileNotification)
            {
                tileNotification->Tag = "tag01";
                TileUpdateManager::CreateTileUpdaterForApplication()->Update(tileNotification);
            }
        }
    }
}

void PrintBackgroundTask::UpdateBadge()
{
    // Get the template
    XmlDocument^ badgeXml = BadgeUpdateManager::GetTemplateContent(BadgeTemplateType::BadgeGlyph);
    if (nullptr != badgeXml)
    {
        XmlElement^ badgeElement = safe_cast<XmlElement^>(badgeXml->SelectSingleNode("/badge"));
        if (nullptr != badgeElement)
        {
            // Fill in the template
            badgeElement->SetAttribute("value", "error");
            BadgeNotification^ badgeNotification = ref new BadgeNotification(badgeXml);
            if (nullptr != badgeNotification)
            {
                // Create a badge update
                BadgeUpdateManager::CreateBadgeUpdaterForApplication()->Update(badgeNotification);
            }
        }
    }
}
